### ISSUE TITLE
[TW-1616] Intermittent lengthy delay when starting or completing a task

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#1640]
_Jeremy John Reeder on 2015-05-13T13:03:01Z says:_ 

 I've noticed lately that there's often a delay of 10 to 20 seconds when I start or complete a task. Here's an example from this morning:
{code}
jeremy(at)fulcrum ~> task 599 start
Starting task 599 'Do 172 squats'.
<INSERT LONG PAUSE HERE>
Started 1 task.
You have more urgent tasks.
Project 'sanec.fortec' is 20% complete (480 of 600 tasks remaining).
jeremy(at)fulcrum ~> task 599 done
Completed task 599 'Do 172 squats'.
Unblocked 600 'Do 172 squats'.
<INSERT LONG PAUSE HERE>
Completed 1 task.
You have more urgent tasks.
The project 'sanec.fortec' has changed.  Project 'sanec.fortec' is 20% complete (479 of 600 tasks remaining).
jeremy(at)fulcrum ~> 
{code}

**Comment by taskwarrior** [comment]
_Jeremy John Reeder on 2015-05-13T18:26:59Z says:_ 

 I know nothing about hook scripts, which I _think_ means that I'm not using any.

**Comment by taskwarrior** [comment]
_Jeremy John Reeder on 2015-05-14T13:11:41Z says:_ 

 Okay, I added `debug=1` to my `.taskrc`. Here's one:

{code}
jeremy(at)fulcrum ~> task 499 start
Starting task 499 'Do 107 pushups'.
Started 1 task.
<INSERT LONG PAUSE HERE>
Timer Config::load (/usr/share/task/rc/holidays.en-US.rc) 0.000701 sec
Timer Config::load (/home/jeremy/.task/holidays) 0.000109 sec
Timer Config::load (/home/jeremy/.task/theme) 0.000329 sec
Timer Config::load (/home/jeremy/.task/reports) 0.000684 sec
Timer Config::load (/home/jeremy/.task/fields) 0.000170 sec
Timer Config::load (/home/jeremy/.taskrc) 0.004889 sec
Hook directory not readable: /home/jeremy/.task/hooks
Filter::subset
  _original_args
    task 499 start
  _args
    arg basename='task' raw='task' ORIGINAL BINARY TW
    argSeq raw='(' FILTER OP
    argSeq raw='(' FILTER OP
    argSeq raw='id' FILTER ATTRIBUTE
    argSeq raw='==' FILTER OP
    argSeq raw='499' FILTER LITERAL NUMBER
    argSeq raw=')' FILTER OP
    argSeq raw=')' FILTER OP
    arg canonical='start' raw='start' ORIGINAL CMD WRITECMD

UUID/ID lexeme found 'id', not applying context.
Derived filter: '( ( ( id == 499 ) ) )'
Filtered 2927 tasks --> 1 tasks [all tasks]
CLI Parser
  _original_args
    task 499 start
  _args
    arg basename='task' raw='task' ORIGINAL BINARY TW
    argSeq raw='(' FILTER OP
    argSeq raw='(' FILTER OP
    argSeq raw='id' FILTER ATTRIBUTE
    argSeq raw='==' FILTER OP
    argSeq raw='499' FILTER LITERAL NUMBER
    argSeq raw=')' FILTER OP
    argSeq raw=')' FILTER OP
    arg canonical='start' raw='start' ORIGINAL CMD WRITECMD

Task::modify
  pending.data rw O T0850+000~001 L0850+000
completed.data rw - T2077+000~000 L2077+000
     undo.data rw O T0000+000~000 L0004+004
  backlog.data rw O T0000+000~000 L0001+001
 
Perf task 2.4.2 48e9c05 20150514T131555Z init:10823 load:317931 gc:0 filter:31969 commit:28859 sort:0 render:0 hooks:7 total:389589

Project 'sanec.fortec' is 20% complete (478 of 600 tasks remaining).
jeremy(at)fulcrum ~> 
{code}

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2015-08-28T22:49:46Z says:_ 

 I'm close to calling this resolved.  Here are the reasons:

1) A bug was fixed in Filter::pendingOnly, which determines whether Taskwarrior needs to load the `completed.data` file at all.  It's a heuristic based on the filter, and was broken, and is now better.  Here is my task database, with 4221 tasks (more than your 3838 tasks):

```
$ task 1 done rc.debug=1
Completed task 1 'Add support for user-specified mappings, per TERM'.
Completed 1 task.
Timer Config::load (/Users/paul/gbf/task-2.4.5.git/doc/rc/dark-gray-blue-256.theme) 0.000404 sec
Timer Config::load (/Users/paul/gbf/task-2.4.5.git/doc/rc/holidays.en-US.rc) 0.000530 sec
Timer Config::load (/Users/paul/.taskrc) 0.004360 sec
Found hook script /Users/paul/.task/hooks/on-add
Found hook script /Users/paul/.task/hooks/on-add-bug-link.py
Found hook script /Users/paul/.task/hooks/on-add.the
Found hook script /Users/paul/.task/hooks/on-exit
Found hook script /Users/paul/.task/hooks/on-launch
Found hook script /Users/paul/.task/hooks/on-launch-good-env
Found hook script /Users/paul/.task/hooks/on-modify
Found hook script /Users/paul/.task/hooks/on-modify.timetracking
Found misnamed hook script /Users/paul/.task/hooks/on-x
Parse Tree (before command-specifіc processing)
  _original_args
    task 1 done rc.debug=1
  _args
    word basename='task' raw='task' BINARY
    number raw='1'
    identifier canonical='done' raw='done' CMD USESCONTEXT ALLOWSFILTER ALLOWSMODIFICATIONS
    pair modifier='debug' name='rc' raw='rc.debug=1' separator='=' value='1' CONFIG

No context.
Filtered 160 tasks --> 1 tasks [pending only]
Task::modify
  pending.data rw O T0160+000~001 L0160+000
completed.data rw - T4061+000~000 L4061+000
     undo.data rw O T0000+000~000 L0004+004
  backlog.data rw O T0000+000~000 L0001+001
 
Perf task 2.4.5 b1839ad 20150828T222233Z init:13394 load:99104 gc:0 filter:1952 commit:7019 sort:0 render:0 hooks:574 total:122043
```

Notice this `Filtered 160 tasks --> 1 tasks [pending only]` which shows that it only loaded pending.data, whereas yours shows `Filtered 3838 tasks --> 1 tasks [all tasks]`.  Version 2.4.5 is clearly doing a better job at reducing the amount of data it *reads*.  This fix will improve your situation, but likely not solve it.

2) The overall performance data for you is this:
```
Perf task 2.4.4 e989dee 20150803T140204Z init:8180 load:573658 gc:0 filter:90467 commit:34531 sort:0 render:0 hooks:8 total:706844
```

For me it is this:
```
Perf task 2.4.5 b1839ad 20150828T222233Z init:13394 load:99104 gc:0 filter:1952 commit:7019 sort:0 render:0 hooks:574 total:122043
```

* The `init` time in 2.4.5 is slower because of the more involved parser.  Not a factor.
* Your `load` time is 573658/99104=5.8 times slower, but then your 2.4.4 loaded 3838/160=24 times more tasks due to (1).  Not a factor.
* Your `filter` time is 90467/1952=46 times slower, due to 24 times more tasks loaded.  Not a factor.
* Your `commit` time (where the files get written) is 34ms, to my 7ms.  Your pending.data file has 1146 tasks to my 160, 7 times more, and your commit time is 5 times longer.  Not a factor.
* Your `hooks` time shows no hooks, and mine take half a millisecond.  Not a factor.
* There is no significant difference in performance here, as recorded internally by Taskwarrior.  The commit time includes processing where the files actually get written and closed.

Summary: The Taskwarrior performance you and I both experience is the same.  The overall time for you is at 10-15 times longer, which I do not see.

I suspect you will see longer times than me, when doing this:

```
$ time task 1 done rc.debug=1
...
Perf task 2.4.5 b1839ad 20150828T225138Z init:6934 load:106373 gc:0 filter:927 commit:2755 sort:0 render:0 hooks:103 total:117092
...
real	0m0.188s
user	0m0.153s
sys	0m0.016s
```

because I see 188ms reported as "real" time, and Taskwarrior internally reports 117ms.  The difference could easily be loading Taskwarrior (and GnuTLS) into memory - launch time, which Taskwarrior cannot record.

If you can run this, and see lengthy delays outside the Taskwarrior runtime, it will point to something about your hardware/software, or maybe Dropbox.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2015-09-12T14:38:49Z says:_ 

 I am now able to reliably replicate this.  Investigating.

```
$ time ./run_perf
Performance: benchmarks
*  task next...
Perf task 2.5.0.beta1 f42de55 20150912T144707Z init:2020 load:82394 gc:29730 filter:19697 commit:2 sort:4085 render:421 hooks:0 total:138349
*  task list...
Perf task 2.5.0.beta1 f42de55 20150912T144708Z init:2000 load:83364 gc:31483 filter:20975 commit:3 sort:24462 render:101642 hooks:0 total:263929
*  task all...
Perf task 2.5.0.beta1 f42de55 20150912T144710Z init:1943 load:86094 gc:31189 filter:71 commit:2 sort:119282 render:502862 hooks:0 total:741443
*  task add...
Perf task 2.5.0.beta1 f42de55 20150912T144711Z init:2108 load:31037 gc:0 filter:0 commit:212 sort:0 render:0 hooks:0 total:33357
*  task export...
Perf task 2.5.0.beta1 f42de55 20150912T144712Z init:1942 load:85442 gc:31667 filter:71 commit:2 sort:0 render:0 hooks:0 total:119124
*  task import...
Perf task 2.5.0.beta1 f42de55 20150912T144911Z init:1947 load:751 gc:0 filter:0 commit:150731 sort:0 render:0 hooks:0 total:153429
End

real	2m5.039s
user	2m0.040s
sys	0m4.863s
```

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2015-09-21T11:32:25Z says:_ 

 We believe this is now fixed.  Here are the reasons:

1. The Filter::pendingOnly method is now properly identifying a filtering shortcut, which was adding a lot of time to some queries.

2. The Nibbler object was making many unnecessary copies of data. We saw the `import` command take minutes to run, because of this, and the Nibbler object is used everywhere.
